### PR TITLE
connection policy: add `local_ip`  matcher

### PR DESCRIPTION
**Pull request by mohammed90**
This is for discussion, because I haven't thoroughly vetted the idea but curious if we can introduce more matchers to behave more like the HTTP handlers matchers.

It might resolve cases like #5771, depending on the use case.

### CODE REVIEW

File: modules/caddytls/matchers.go [hunk 1]
```diff
@@ -144,8 +147,179 @@ func (MatchRemoteIP) matches(ip netip.Addr, ranges []netip.Prefix) bool {
...
+func (mc MatchNot) Match(hello *tls.ClientHelloInfo) bool {
+	for _, ms := range mc.matchers {
+		if ms.Match(hello) {
+			return false
+		}
+	}
+	return true
+}
+
 // Interface guards
```
[the remark this answers is not in the document]
Comment (mohammed90) [reply modules/caddytls/matchers.go:320]: Done

File: modules/caddytls/matchers.go [hunk 2]
```diff
@@ -144,8 +147,179 @@ func (MatchRemoteIP) matches(ip netip.Addr, ranges []netip.Prefix) bool {
...
+func (ms MatcherSets) AnyMatch(hello *tls.ClientHelloInfo) bool {
+	for _, mset := range ms {
+		if mset.Match(hello) {
+			return true
+		}
+	}
+	return false
+}
+
+type MatchNot struct {
```
[the remark this answers is not in the document]
Comment (mohammed90) [reply modules/caddytls/matchers.go]: Done

**Comment by mohammed90** [comment]
Changing the PR to draft until I add tests and we have consensus on the discussion whether this is a good change

**Comment by mholt** [comment]
I like where this is going.

What if we just start with the local IP matcher and then see about adding the complexity of the matcher sets later if we feel it is needed based on real-world feedback?

**Comment by mohammed90** [comment]
> What if we just start with the local IP matcher and then see about adding the complexity of the matcher sets later if we feel it is needed based on real-world feedback?

Done.

File: modules/caddytls/matchers.go [hunk 3]
```diff
@@ -65,6 +66,7 @@ type MatchRemoteIP struct {
 	Ranges []string `json:"ranges,omitempty"`
 
 	// The IPs or CIDR ranges to *NOT* match.
+	// Deprecated: Use `not` matcher instead.
```
Comment (mholt) [inline modules/caddytls/matchers.go:69]:
```suggestion
```

For now :)

**Review by mholt — APPROVED** [review]
Cool, thank you!

I wonder if we should use this in the Caddyfile parser when crafting TLS connection policies -- instead of always using ServerName, which will be empty then the host is the IP address.

Not for this PR, but just thinking for future improvement.

**Pull request merged**